File: binding/src/libqpdfwriter.cpp
#include <QByteArray>
#include <QIODevice>
#include <QObject>
#include <QPagedPaintDevice>
#include <QPaintEngine>
#include <QPdfOutputIntent>
#include <QPdfWriter>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QUuid>
#include <qpdfwriter.h>
#include "libqpdfwriter.h"
#include "libqpdfwriter.hxx"

QPdfWriter* QPdfWriter_new(const libqt_string filename) {
	 return new VirtualQPdfWriter(QString::fromUtf8(filename.data, filename.len));
}

QPdfWriter* QPdfWriter_new2(QIODevice* device) {
	 return new VirtualQPdfWriter(device);
}

libqt_string QPdfWriter_Tr(const char* s) {
	QString _ret = QPdfWriter::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPdfWriter_SetPdfVersion(QPdfWriter* self, int version) {
	self->setPdfVersion(static_cast<QPagedPaintDevice::PdfVersion>(version));
}

int QPdfWriter_PdfVersion(const QPdfWriter* self) {
	return self->pdfVersion();
}

libqt_string QPdfWriter_Title(const QPdfWriter* self) {
	QString _ret = self->title();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPdfWriter_SetTitle(QPdfWriter* self, const libqt_string title) {
	self->setTitle(QString::fromUtf8(title.data, title.len));
}

libqt_string QPdfWriter_Creator(const QPdfWriter* self) {
	QString _ret = self->creator();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPdfWriter_SetCreator(QPdfWriter* self, const libqt_string creator) {
	self->setCreator(QString::fromUtf8(creator.data, creator.len));
}

QUuid* QPdfWriter_DocumentId(const QPdfWriter* self) {
	return new QUuid(self->documentId());
}

void QPdfWriter_SetDocumentId(QPdfWriter* self, QUuid* documentId) {
	self->setDocumentId(*documentId);
}

bool QPdfWriter_NewPage(QPdfWriter* self) {
	return self->newPage();
}

void QPdfWriter_SetResolution(QPdfWriter* self, int resolution) {
	self->setResolution(resolution);
}

int QPdfWriter_Resolution(const QPdfWriter* self) {
	return self->resolution();
}

void QPdfWriter_SetDocumentXmpMetadata(QPdfWriter* self, const libqt_string xmpMetadata) {
	self->setDocumentXmpMetadata(QByteArray(xmpMetadata.data, xmpMetadata.len));
}

libqt_string QPdfWriter_DocumentXmpMetadata(const QPdfWriter* self) {
	QByteArray _qb = self->documentXmpMetadata();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPdfWriter_AddFileAttachment(QPdfWriter* self, const libqt_string fileName, const libqt_string data) {
	self->addFileAttachment(QString::fromUtf8(fileName.data, fileName.len), QByteArray(data.data, data.len));
}

int QPdfWriter_ColorModel(const QPdfWriter* self) {
	return self->colorModel();
}

void QPdfWriter_SetColorModel(QPdfWriter* self, int model) {
	self->setColorModel(static_cast<QPixelFormat::ColorModel>(model));
}

QPdfOutputIntent* QPdfWriter_OutputIntent(const QPdfWriter* self) {
	return new QPdfOutputIntent(self->outputIntent());
}

void QPdfWriter_SetOutputIntent(QPdfWriter* self, const QPdfOutputIntent* intent) {
	self->setOutputIntent(*intent);
}

libqt_string QPdfWriter_Tr2(const char* s, const char* c) {
	QString _ret = QPdfWriter::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QPdfWriter_Tr3(const char* s, const char* c, int n) {
	QString _ret = QPdfWriter::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPdfWriter_AddFileAttachment3(QPdfWriter* self, const libqt_string fileName, const libqt_string data, const libqt_string mimeType) {
	self->addFileAttachment(QString::fromUtf8(fileName.data, fileName.len), QByteArray(data.data, data.len), QString::fromUtf8(mimeType.data, mimeType.len));
}

// Base class handler implementation
bool QPdfWriter_QBaseNewPage(QPdfWriter* self) {
	auto* vqpdfwriter = dynamic_cast<VirtualQPdfWriter*>(self);
	if (vqpdfwriter && vqpdfwriter->isVirtualQPdfWriter) {
vqpdfwriter->setQPdfWriter_NewPage_IsBase(true);
	return vqpdfwriter->newPage();
}
}

// Auxiliary method to allow providing re-implementation
void QPdfWriter_OnNewPage(QPdfWriter* self, intptr_t slot) {
	auto* vqpdfwriter = dynamic_cast<VirtualQPdfWriter*>(self);
	if (vqpdfwriter && vqpdfwriter->isVirtualQPdfWriter) {
vqpdfwriter->setQPdfWriter_NewPage_Callback(reinterpret_cast<VirtualQPdfWriter::QPdfWriter_NewPage_Callback>(slot));
}
}

void QPdfWriter_Delete(QPdfWriter* self) {
    delete self;
}

File: binding/src/libqpdfwriter.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qpdfwriter.html
pub const qpdfwriter = struct {

    /// New constructs a new QPdfWriter object.
    pub fn New(filename: []const u8) QtC.QPdfWriter {
        const filename_str = qtc.libqt_string{
    .len = filename.len,
    .data = filename.ptr,
};
return qtc.QPdfWriter_new(filename_str);
    }


    /// New2 constructs a new QPdfWriter object.
    pub fn New2(device: ?*anyopaque) QtC.QPdfWriter {
        return qtc.QPdfWriter_new2(@ptrCast(device));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPdfWriter_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpdfwriter.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetPdfVersion(self: ?*anyopaque, version: i32) void {
        qtc.QPdfWriter_SetPdfVersion(@ptrCast(self), @intCast(version));
    }

    pub fn PdfVersion(self: ?*anyopaque, ) i32 {
        return qtc.QPdfWriter_PdfVersion(@ptrCast(self));
    }

    pub fn Title(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPdfWriter_Title(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpdfwriter.Title: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetTitle(self: ?*anyopaque, title: []const u8) void {
        const title_str = qtc.libqt_string{
    .len = title.len,
    .data = title.ptr,
};
qtc.QPdfWriter_SetTitle(@ptrCast(self), title_str);
    }

    pub fn Creator(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPdfWriter_Creator(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpdfwriter.Creator: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetCreator(self: ?*anyopaque, creator: []const u8) void {
        const creator_str = qtc.libqt_string{
    .len = creator.len,
    .data = creator.ptr,
};
qtc.QPdfWriter_SetCreator(@ptrCast(self), creator_str);
    }

    pub fn DocumentId(self: ?*anyopaque, ) QtC.QUuid {
        return qtc.QPdfWriter_DocumentId(@ptrCast(self));
    }

    pub fn SetDocumentId(self: ?*anyopaque, documentId: QtC.QUuid) void {
        qtc.QPdfWriter_SetDocumentId(@ptrCast(self), @ptrCast(documentId));
    }

    pub fn NewPage(self: ?*anyopaque, ) bool {
        return qtc.QPdfWriter_NewPage(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnNewPage(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) bool) void {
        qtc.QPdfWriter_OnNewPage(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseNewPage(self: ?*anyopaque, ) bool {
        return qtc.QPdfWriter_QBaseNewPage(@ptrCast(self));
    }

    pub fn SetResolution(self: ?*anyopaque, resolution: i32) void {
        qtc.QPdfWriter_SetResolution(@ptrCast(self), resolution);
    }

    pub fn Resolution(self: ?*anyopaque, ) i32 {
        return qtc.QPdfWriter_Resolution(@ptrCast(self));
    }

    pub fn SetDocumentXmpMetadata(self: ?*anyopaque, xmpMetadata: []u8) void {
        const xmpMetadata_str = qtc.libqt_string{
    .len = xmpMetadata.len,
    .data = xmpMetadata.ptr,
};
qtc.QPdfWriter_SetDocumentXmpMetadata(@ptrCast(self), xmpMetadata_str);
    }

    pub fn DocumentXmpMetadata(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QPdfWriter_DocumentXmpMetadata(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qpdfwriter.DocumentXmpMetadata: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn AddFileAttachment(self: ?*anyopaque, fileName: []const u8, data: []u8) void {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
const data_str = qtc.libqt_string{
    .len = data.len,
    .data = data.ptr,
};
qtc.QPdfWriter_AddFileAttachment(@ptrCast(self), fileName_str, data_str);
    }

    pub fn ColorModel(self: ?*anyopaque, ) i32 {
        return qtc.QPdfWriter_ColorModel(@ptrCast(self));
    }

    pub fn SetColorModel(self: ?*anyopaque, model: i32) void {
        qtc.QPdfWriter_SetColorModel(@ptrCast(self), @intCast(model));
    }

    pub fn OutputIntent(self: ?*anyopaque, ) QtC.QPdfOutputIntent {
        return qtc.QPdfWriter_OutputIntent(@ptrCast(self));
    }

    pub fn SetOutputIntent(self: ?*anyopaque, intent: ?*anyopaque) void {
        qtc.QPdfWriter_SetOutputIntent(@ptrCast(self), @ptrCast(intent));
    }

    pub fn PaintEngine(self: ?*anyopaque, ) QtC.QPaintEngine {
        return qtc.QPdfWriter_PaintEngine(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnPaintEngine(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QPaintEngine) void {
        qtc.QPdfWriter_OnPaintEngine(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBasePaintEngine(self: ?*anyopaque, ) QtC.QPaintEngine {
        return qtc.QPdfWriter_QBasePaintEngine(@ptrCast(self));
    }

    pub fn Metric(self: ?*anyopaque, id: i32) i32 {
        return qtc.QPdfWriter_Metric(@ptrCast(self), @intCast(id));
    }

    /// Allows for overriding the related default method
    pub fn OnMetric(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) i32) void {
        qtc.QPdfWriter_OnMetric(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMetric(self: ?*anyopaque, id: i32) i32 {
        return qtc.QPdfWriter_QBaseMetric(@ptrCast(self), @intCast(id));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPdfWriter_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpdfwriter.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPdfWriter_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpdfwriter.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn AddFileAttachment3(self: ?*anyopaque, fileName: []const u8, data: []u8, mimeType: []const u8) void {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
const data_str = qtc.libqt_string{
    .len = data.len,
    .data = data.ptr,
};
const mimeType_str = qtc.libqt_string{
    .len = mimeType.len,
    .data = mimeType.ptr,
};
qtc.QPdfWriter_AddFileAttachment3(@ptrCast(self), fileName_str, data_str, mimeType_str);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPdfWriter_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/pdfwriter.html#types
pub const enums = struct {
    pub const ColorModel = enum {
        pub const RGB: i32 = 0;
        pub const Grayscale: i32 = 1;
        pub const CMYK: i32 = 2;
        pub const Auto: i32 = 3;
    };

};
